{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TODO:\n",
    "-  ADD the training sequences (see how to get it on : https://github.com/pinellolab/DNA-Diffusion/blob/main/src/dnadiffusion/data/K562_hESCT0_HepG2_GM12878_12k_sequences_per_group.txt). Capture 200 nucleotide regions from the center of these coordinates (check bedtools)\n",
    "-  Lets try to have table with all the endogenous sequences, but for NOW just 1k regions from the training set\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tensorflow_hub as hub\n",
    "import joblib\n",
    "import gzip\n",
    "import kipoiseq\n",
    "from kipoiseq import Interval\n",
    "import pyfaidx\n",
    "import pyBigWig\n",
    "import igv_notebook\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "import os \n",
    "import pickle as pkl\n",
    "import wget \n",
    "from tqdm import tqdm_notebook as tqdm\n",
    "import tensorflow as tf\n",
    "from enformer_dna_diff.enformer import Enformer, FastaStringExtractor\n",
    "from enformer_dna_diff.enformerops import EnformerOps, SEQ_EXTRACT\n",
    "\n",
    "%matplotlib inline\n",
    "%config InlineBackend.figure_format = 'retina'\n",
    "\n",
    "# assert tf.config.list_physical_devices('GPU'), 'Start the colab kernel with GPU: Runtime -> Change runtime type -> GPU'\n",
    "\n",
    "# tf.config.list_physical_devices('GPU')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5_31_2023_enformer_CAGE_DNAse_bias_experiment.ipynb\r\n",
      "CAGE:GM12878_enformer.bw\r\n",
      "CAGE:HepG2_enformer.bw\r\n",
      "CAGE:K562_enformer.bw\r\n",
      "DNASE:GM12878_enformer.bw\r\n",
      "DNASE:H1esc_enformer.bw\r\n",
      "DNASE:HepG2_enformer.bw\r\n",
      "DNASE:K562_enformer.bw\r\n",
      "DNASE_RANDOM_SEQS.TXT\r\n",
      "DNA_DIFFUSION_ALL_SEQS.txt\r\n",
      "Enformer_Experiments\r\n",
      "GENERATED_SEQS.TXT\r\n",
      "MASTER_DNA_DIFFUSION_ALL_SEQS.txt\r\n",
      "PROMOTERS_SEQS.TXT\r\n",
      "TRAINING_TEST_VALIDATION_SEQS.TXT\r\n",
      "Untitled.ipynb\r\n",
      "bin\r\n",
      "boot\r\n",
      "data\r\n",
      "demo_out.ipynb\r\n",
      "demo_out_tq.ipynb\r\n",
      "dev\r\n",
      "etc\r\n",
      "hg38.chrom.sizes\r\n",
      "hg38_liftover+new_CAGE_peaks_phase1and2.bed.gz\r\n",
      "home\r\n",
      "lib\r\n",
      "lib64\r\n",
      "media\r\n",
      "mnt\r\n",
      "opt\r\n",
      "proc\r\n",
      "root\r\n",
      "run\r\n",
      "sbin\r\n",
      "srv\r\n",
      "sys\r\n",
      "tmp\r\n",
      "tqdm.ipynb\r\n",
      "usr\r\n",
      "var\r\n"
     ]
    }
   ],
   "source": [
    "!ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "model_path = 'https://tfhub.dev/deepmind/enformer/1'\n",
    "fasta_file = 'data/genome.fa'\n",
    "targets_txt = 'https://raw.githubusercontent.com/calico/basenji/master/manuscripts/cross2020/targets_human.txt'\n",
    "df_targets = pd.read_csv(targets_txt, sep='\\t')\n",
    "df_targets[df_targets.apply(lambda x : '' in x['description'] and  'CAGE' in x['description'] , 1)]\n",
    "df_sizes = pd.read_table('data/hg38.chrom.sizes', header=None).head(22)\n",
    "SEQUENCE_LENGTH = 393216\n",
    "model = Enformer(model_path)\n",
    "fasta_extractor = FastaStringExtractor(fasta_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ID</th>\n",
       "      <th>SEQUENCE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAG</th>\n",
       "      <th>CELL_TYPE</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">GENERATED</th>\n",
       "      <th>GM12878</th>\n",
       "      <td>100000</td>\n",
       "      <td>100000</td>\n",
       "      <td>100000</td>\n",
       "      <td>100000</td>\n",
       "      <td>100000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HEPG2</th>\n",
       "      <td>100000</td>\n",
       "      <td>100000</td>\n",
       "      <td>100000</td>\n",
       "      <td>100000</td>\n",
       "      <td>100000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HESCT0</th>\n",
       "      <td>100000</td>\n",
       "      <td>100000</td>\n",
       "      <td>100000</td>\n",
       "      <td>100000</td>\n",
       "      <td>100000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K562</th>\n",
       "      <td>100000</td>\n",
       "      <td>100000</td>\n",
       "      <td>100000</td>\n",
       "      <td>100000</td>\n",
       "      <td>100000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PROMOTERS</th>\n",
       "      <th>NO</th>\n",
       "      <td>20013</td>\n",
       "      <td>20013</td>\n",
       "      <td>20013</td>\n",
       "      <td>20013</td>\n",
       "      <td>20013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RANDOM_GENOME_REGIONS</th>\n",
       "      <th>NO</th>\n",
       "      <td>5000</td>\n",
       "      <td>5000</td>\n",
       "      <td>5000</td>\n",
       "      <td>5000</td>\n",
       "      <td>5000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">test</th>\n",
       "      <th>GM12878</th>\n",
       "      <td>1044</td>\n",
       "      <td>1044</td>\n",
       "      <td>1044</td>\n",
       "      <td>1044</td>\n",
       "      <td>1044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HepG2</th>\n",
       "      <td>958</td>\n",
       "      <td>958</td>\n",
       "      <td>958</td>\n",
       "      <td>958</td>\n",
       "      <td>958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K562</th>\n",
       "      <td>1321</td>\n",
       "      <td>1321</td>\n",
       "      <td>1321</td>\n",
       "      <td>1321</td>\n",
       "      <td>1321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hESCT0</th>\n",
       "      <td>1242</td>\n",
       "      <td>1242</td>\n",
       "      <td>1242</td>\n",
       "      <td>1242</td>\n",
       "      <td>1242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">training</th>\n",
       "      <th>GM12878</th>\n",
       "      <td>9903</td>\n",
       "      <td>9903</td>\n",
       "      <td>9903</td>\n",
       "      <td>9903</td>\n",
       "      <td>9903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HepG2</th>\n",
       "      <td>9833</td>\n",
       "      <td>9833</td>\n",
       "      <td>9833</td>\n",
       "      <td>9833</td>\n",
       "      <td>9833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K562</th>\n",
       "      <td>9769</td>\n",
       "      <td>9769</td>\n",
       "      <td>9769</td>\n",
       "      <td>9769</td>\n",
       "      <td>9769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hESCT0</th>\n",
       "      <td>9749</td>\n",
       "      <td>9749</td>\n",
       "      <td>9749</td>\n",
       "      <td>9749</td>\n",
       "      <td>9749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">validation</th>\n",
       "      <th>GM12878</th>\n",
       "      <td>1021</td>\n",
       "      <td>1021</td>\n",
       "      <td>1021</td>\n",
       "      <td>1021</td>\n",
       "      <td>1021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HepG2</th>\n",
       "      <td>1177</td>\n",
       "      <td>1177</td>\n",
       "      <td>1177</td>\n",
       "      <td>1177</td>\n",
       "      <td>1177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K562</th>\n",
       "      <td>878</td>\n",
       "      <td>878</td>\n",
       "      <td>878</td>\n",
       "      <td>878</td>\n",
       "      <td>878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hESCT0</th>\n",
       "      <td>977</td>\n",
       "      <td>977</td>\n",
       "      <td>977</td>\n",
       "      <td>977</td>\n",
       "      <td>977</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  chrom   start     end      ID  SEQUENCE\n",
       "TAG                   CELL_TYPE                                          \n",
       "GENERATED             GM12878    100000  100000  100000  100000    100000\n",
       "                      HEPG2      100000  100000  100000  100000    100000\n",
       "                      HESCT0     100000  100000  100000  100000    100000\n",
       "                      K562       100000  100000  100000  100000    100000\n",
       "PROMOTERS             NO          20013   20013   20013   20013     20013\n",
       "RANDOM_GENOME_REGIONS NO           5000    5000    5000    5000      5000\n",
       "test                  GM12878      1044    1044    1044    1044      1044\n",
       "                      HepG2         958     958     958     958       958\n",
       "                      K562         1321    1321    1321    1321      1321\n",
       "                      hESCT0       1242    1242    1242    1242      1242\n",
       "training              GM12878      9903    9903    9903    9903      9903\n",
       "                      HepG2        9833    9833    9833    9833      9833\n",
       "                      K562         9769    9769    9769    9769      9769\n",
       "                      hESCT0       9749    9749    9749    9749      9749\n",
       "validation            GM12878      1021    1021    1021    1021      1021\n",
       "                      HepG2        1177    1177    1177    1177      1177\n",
       "                      K562          878     878     878     878       878\n",
       "                      hESCT0        977     977     977     977       977"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "Data structure"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "SEQS = SEQ_EXTRACT('data/MASTER_DNA_DIFFUSION_ALL_SEQS.txt')\n",
    "SEQS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -Add the random for our full dataset.\n",
    "# -Dont remove hesct0 (use h1esc) use NA for cage\n",
    "# -Generate 10k random sequences (remove regions with NN)\n",
    "# -Add the CAGE tss(1kb+/-) prediction for the other 3 celltypes on the 440k.\n",
    "# -Dont show tss in the final sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# eops.add_track({\n",
    "#     'name': 'DNAse:hESC_enformer',\n",
    "#     'file': None,\n",
    "#     'color': 'BLUE',\n",
    "#     'type': 'enformer',\n",
    "#     'id': 123,\n",
    "#     'log': False\n",
    "# })\n",
    "\n",
    "def load_tracks_eops(eops):\n",
    "    \n",
    "    eops.add_track({\n",
    "        'name': 'DNASE:GM12878_enformer',\n",
    "        'file': None,\n",
    "        'color': 'BLUE',\n",
    "        'type': 'enformer',\n",
    "        'id': 12,\n",
    "        'log': False\n",
    "    })\n",
    "\n",
    "    eops.add_track({\n",
    "        'name': 'DNASE:K562_enformer',\n",
    "        'file': None,\n",
    "        'color': 'BLUE',\n",
    "        'type': 'enformer',\n",
    "        'id': 121,\n",
    "        'log': False\n",
    "    })\n",
    "\n",
    "    eops.add_track({\n",
    "        'name': 'DNASE:HepG2_enformer',\n",
    "        'file': None,\n",
    "        'color': 'BLUE',\n",
    "        'type': 'enformer',\n",
    "        'id': 27,\n",
    "        'log': False\n",
    "    })\n",
    "    \n",
    "    eops.add_track({\n",
    "        'name': 'DNASE:H1esc_enformer',\n",
    "        'file': None,\n",
    "        'color': 'BLUE',\n",
    "        'type': 'enformer',\n",
    "        'id': 19,\n",
    "        'log': False\n",
    "    })\n",
    "\n",
    "    eops.add_track({\n",
    "        'name': 'CAGE:K562_enformer',\n",
    "        'file': None,\n",
    "        'color': 'RED',\n",
    "        'type': 'enformer',\n",
    "        'id': 5111,\n",
    "        'log': True\n",
    "    })\n",
    "\n",
    "    eops.add_track({\n",
    "        'name': 'CAGE:GM12878_enformer',\n",
    "        'file': None,\n",
    "        'color': 'RED',\n",
    "        'type': 'enformer',\n",
    "        'id': 5110,\n",
    "        'log': True\n",
    "    })\n",
    "\n",
    "    eops.add_track({\n",
    "        'name': 'CAGE:HepG2_enformer',\n",
    "        'file': None,\n",
    "        'color': 'RED',\n",
    "        'type': 'enformer',\n",
    "        'id': 5109,\n",
    "        'log': True\n",
    "    })\n",
    "    \n",
    "    # eops.add_track({\n",
    "    #     'name': 'CAGE:hESC_enformer',\n",
    "    #     'file': None,\n",
    "    #     'color': 'RED',\n",
    "    #     'type': 'enformer',\n",
    "    #     'id': 500,\n",
    "    #     'log': True\n",
    "    # })\n",
    "\n",
    "    # eops.add_track({\n",
    "    #     'name': 'K562 DNASE REAL',\n",
    "    #     'file': 'ENCFF413AHU.bigWig',\n",
    "    #     'color': 'skyblue',\n",
    "    #     'type': 'real'\n",
    "    # })\n",
    "    # eops.add_track({\n",
    "    #     'name': 'GM12878 DNASE REAL',\n",
    "    #     'file': 'ENCFF093VXI.bigWig',\n",
    "    #     'color': 'skyblue',\n",
    "    #     'type': 'real'\n",
    "    # })\n",
    "\n",
    "    # eops.remove_track(['CAGE:hESC_enformer', 'DNAse:hESC_enformer'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Enformer output analysis of training; random; and generated sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# all_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6 hours  random\n",
    "# 20 hours promoters \n",
    "# 50 hours train/test/validation\n",
    "# 350 hours  GENERATED"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DEMO=True\n",
    "all_data = SEQS.data.copy()\n",
    "MODIFY_PREFIX= '1_'\n",
    "#run lucas_test\n",
    "#papermill  5_31_2023_enformer_CAGE_DNAse_bias_experiment.ipynb  demo.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Random sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Remove all head 3\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['RANDOM_GENOME_REGIONS'] (5000, 7)\n"
     ]
    }
   ],
   "source": [
    "print('Remove all head 3')\n",
    "eops = EnformerOps()\n",
    "eops.load_data(SEQS.extract_seq('GENERATED', 'GM12878')['SEQUENCE'].values.tolist())\n",
    "load_tracks_eops(eops) #loading the tracks\n",
    "\n",
    "ran_seqs = all_data[all_data['TAG'] == 'RANDOM_GENOME_REGIONS'].copy()\n",
    "print (ran_seqs['TAG'].unique(), ran_seqs.shape)\n",
    "\n",
    "subset_ran_seqs = ran_seqs[['chrom', 'start', 'end', 'ID']]\n",
    "#ran_seq_list = subset_ran_seqs.head(3).values.tolist() # removing demo\n",
    "\n",
    "ran_seq_list = subset_ran_seqs.values.tolist() # removing demo\n",
    "if DEMO:\n",
    "    ran_seq_list=ran_seq_list[:3]\n",
    "\n",
    "captured_values = []\n",
    "for s in ran_seq_list: \n",
    "    #print(s)\n",
    "    try:\n",
    "        s_in = [s[0], int(s[1]), int(s[2])]\n",
    "        id_seq = s[3] \n",
    "        list_bw = eops.generate_plot_number(0, interval_list=s_in, wildtype=True, \n",
    "                                          show_track=False, modify_prefix=MODIFY_PREFIX) \n",
    "    except RuntimeError:\n",
    "      # Infrequent \"the entries are out of order\" error for some random seqs \n",
    "        continue\n",
    "    try:\n",
    "        out_in = eops.extract_from_position(s_in, as_dataframe=True)\n",
    "        #print (out_in.shape)\n",
    "        out_in = out_in.mean()\n",
    "        out_in['SEQ_ID'] = id_seq\n",
    "        out_in['TARGET_NAME'] = 'ITSELF'\n",
    "        columns_names = out_in.copy()\n",
    "        captured_values.append( out_in  )\n",
    "    except ValueError:\n",
    "      # Infrequent \"All arrays must be of the same length\" error\n",
    "        continue\n",
    "\n",
    "df_out = pd.DataFrame([x.values.tolist() for x in captured_values], columns=out_in.index)\n",
    "\n",
    "df_out.to_csv('DNASE_RANDOM_SEQS.TXT', sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Promoters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ['RANDOM_GENOME_REGIONS']        (5000)    (MEASURING: 200bp)       #DNASE (K562,HEPG2,GM, H1ESC) # CAGE (K562,HEPG2,GM)\n",
    "# ['PROMOTERS']                    (20013)  (MEASURING: 2000bp)       #DNASE (K562,HEPG2,GM, H1ESC) # CAGE (K562,HEPG2,GM)\n",
    "# ['training' 'test' 'validation'] (47872)   (MEASURING: 200bp)       #DNASE (K562,HEPG2,GM, H1ESC) # CAGE (K562,HEPG2,GM)\n",
    "\n",
    "# #2X (ENHANCER AND GATA1)  #DNASE (K562,HEPG2,GM, H1ESC) # CAGE (K562,HEPG2,GM)\n",
    "# ['GENERATED']                    (400000)   (MEASURING: ENH(200bp) GATA1_GB( 7925bp) ) \n",
    "                            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Remove all head 3\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['PROMOTERS'] (20013, 7)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:16: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0\n",
      "Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`\n",
      "  app.launch_new_instance()\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "653a8c4c89084430bda686c507a00845",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/3 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print('Remove all head 3')\n",
    "eops = EnformerOps()\n",
    "eops.load_data(SEQS.extract_seq('GENERATED', 'GM12878')['SEQUENCE'].values.tolist())\n",
    "load_tracks_eops(eops) #loading the tracks\n",
    "\n",
    "ran_seqs = all_data[all_data['TAG'] == 'PROMOTERS'].copy()\n",
    "print (ran_seqs['TAG'].unique(), ran_seqs.shape)\n",
    "\n",
    "subset_ran_seqs = ran_seqs[['chrom', 'start', 'end', 'ID']]\n",
    "#ran_seq_list = subset_ran_seqs.head(3).values.tolist()\n",
    "ran_seq_list = subset_ran_seqs.values.tolist()\n",
    "if DEMO:\n",
    "    ran_seq_list=ran_seq_list[:3]\n",
    "\n",
    "captured_values = []\n",
    "for s in tqdm(ran_seq_list): \n",
    "    #print(s)\n",
    "    try:\n",
    "        s_in = [s[0], int(s[1]), int(s[2])]\n",
    "        id_seq = s[3] \n",
    "        list_bw = eops.generate_plot_number(0, interval_list=s_in, wildtype=True, \n",
    "                                          show_track=False, modify_prefix=MODIFY_PREFIX) \n",
    "    except RuntimeError:\n",
    "      # Infrequent \"the entries are out of order\" error for some random seqs \n",
    "        continue\n",
    "    try:\n",
    "        out_in = eops.extract_from_position(s_in, as_dataframe=True)\n",
    "        #print (out_in.shape)\n",
    "        out_in = out_in.mean()\n",
    "        out_in['SEQ_ID'] = id_seq\n",
    "        out_in['TARGET_NAME'] = 'ITSELF'\n",
    "        columns_names = out_in.copy()\n",
    "        captured_values.append( out_in  )\n",
    "    except ValueError:\n",
    "      # Infrequent \"All arrays must be of the same length\" error\n",
    "        continue\n",
    "\n",
    "df_out = pd.DataFrame([x.values.tolist() for x in captured_values], columns=out_in.index)\n",
    "\n",
    "df_out.to_csv('PROMOTERS_SEQS.TXT', sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TRAINING"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Remove all head 3\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['training' 'test' 'validation'] (47872, 7)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:15: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0\n",
      "Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`\n",
      "  from ipykernel import kernelapp as app\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a572071feb474734b14019b88af5dfe4",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/3 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print('Remove all head 3')\n",
    "eops = EnformerOps()\n",
    "eops.load_data(SEQS.extract_seq('GENERATED', 'GM12878')['SEQUENCE'].values.tolist())\n",
    "load_tracks_eops(eops) #loading the tracks\n",
    "\n",
    "ran_seqs = all_data[all_data['TAG'].apply(lambda x : x in {'training', 'test', 'validation'})].copy()\n",
    "print (ran_seqs['TAG'].unique(), ran_seqs.shape)\n",
    "subset_ran_seqs = ran_seqs[['chrom', 'start', 'end', 'ID']]\n",
    "#ran_seq_list = subset_ran_seqs.head(3).values.tolist()\n",
    "ran_seq_list = subset_ran_seqs.values.tolist()\n",
    "if DEMO:\n",
    "    ran_seq_list=ran_seq_list[:3]\n",
    "\n",
    "captured_values = []\n",
    "for s in tqdm(ran_seq_list): \n",
    "    #print(s)\n",
    "    try:\n",
    "        s_in = [s[0], int(s[1]), int(s[2])]\n",
    "        id_seq = s[3] \n",
    "        list_bw = eops.generate_plot_number(0, interval_list=s_in, wildtype=True, \n",
    "                                          show_track=False, modify_prefix=MODIFY_PREFIX) \n",
    "    except RuntimeError:\n",
    "      # Infrequent \"the entries are out of order\" error for some random seqs \n",
    "        continue\n",
    "    try:\n",
    "        out_in = eops.extract_from_position(s_in, as_dataframe=True)\n",
    "        #print (out_in.shape)\n",
    "        out_in = out_in.mean()\n",
    "        out_in['SEQ_ID'] = id_seq\n",
    "        out_in['TARGET_NAME'] = 'ITSELF'\n",
    "        columns_names = out_in.copy()\n",
    "        captured_values.append( out_in  )\n",
    "    except ValueError:\n",
    "      # Infrequent \"All arrays must be of the same length\" error\n",
    "        continue\n",
    "\n",
    "df_out = pd.DataFrame([x.values.tolist() for x in captured_values], columns=out_in.index)\n",
    "\n",
    "df_out.to_csv('TRAINING_TEST_VALIDATION_SEQS.TXT', sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Remove all head 3')\n",
    "eops = EnformerOps()\n",
    "\n",
    "\n",
    "\n",
    "load_tracks_eops(eops) #loading the tracks\n",
    "\n",
    "ran_seqs = all_data[all_data['TAG'] == 'GENERATED'].copy()\n",
    "subset_ran_seqs = ran_seqs[['SEQUENCE', 'ID']]\n",
    "print ( all_data[all_data['TAG'] == 'GENERATED'].shape)\n",
    "print (ran_seqs['TAG'].unique())\n",
    "\n",
    "#ran_seq_list = subset_ran_seqs.head(3).values.tolist()\n",
    "ran_seq_list = subset_ran_seqs.values.tolist()\n",
    "if DEMO:\n",
    "    ran_seq_list=ran_seq_list[:3]\n",
    "\n",
    "ENHANCER_REGION =  ['chrX', 48782929, 48783129]\n",
    "GENE_REGION = ['chrX', 48786486, 48794411]  # 48786486,48794411\n",
    "GENE_NAME = 'GATA1'\n",
    "\n",
    "captured_values = []\n",
    "captured_values_target = []\n",
    "for s in tqdm(ran_seq_list): \n",
    "    \n",
    "    try:\n",
    "        s_in = s[1]\n",
    "        id_seq = s[0]   # aways inject a sequence\n",
    "        #print (id_seq)\n",
    "        eops.load_data([[id_seq]]) # This will be always zero (sequence passed using insert_seq_directly)\n",
    "        list_bw = eops.generate_plot_number(-1, interval_list=ENHANCER_REGION, wildtype=False, \n",
    "                                          show_track=False, modify_prefix=MODIFY_PREFIX) \n",
    "    except RuntimeError:\n",
    "      # Infrequent \"the entries are out of order\" error for some random seqs \n",
    "        continue\n",
    "    try:\n",
    "        out_in = eops.extract_from_position(ENHANCER_REGION, as_dataframe=True)\n",
    "        #print (out_in.shape)\n",
    "        out_in = out_in.mean()\n",
    "        out_in['SEQ_ID'] = id_seq\n",
    "        out_in['TARGET_NAME'] = 'ENH_GATA1'\n",
    "        columns_names = out_in.copy()\n",
    "        captured_values.append( out_in  )\n",
    "        \n",
    "        out_in = eops.extract_from_position(GENE_REGION, as_dataframe=True)\n",
    "        #print (out_in.shape)\n",
    "        out_in = out_in.mean()\n",
    "        out_in['SEQ_ID'] = id_seq\n",
    "        out_in['TARGET_NAME'] = 'GATA1'\n",
    "        columns_names = out_in.copy()\n",
    "        captured_values_target.append( out_in  )\n",
    "        \n",
    "    except ValueError:\n",
    "      # Infrequent \"All arrays must be of the same length\" error\n",
    "        continue\n",
    "\n",
    "df_out_ENH = pd.DataFrame([x.values.tolist() for x in captured_values], columns=['ENHANCER_' + x for x in   out_in.index])\n",
    "df_out_GENE = pd.DataFrame([x.values.tolist() for x in captured_values_target], columns=['GENE_' + x for x in   out_in.index])\n",
    "\n",
    "df_out = pd.concat([df_out_ENH, df_out_GENE], axis=1)\n",
    "\n",
    "df_out.to_csv('GENERATED_SEQS.TXT', sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_out_ENH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_out_GENE"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
